csrc/scheduler/matmul.cpp

```diff
@@ -438,6 +438,89 @@ void scheduleProlog(TensorView* shared_mem_tv, const MatmulParams& params) {
 
 } // namespace
 
+namespace {
+int getKAxis(const MmaLayout& layout) {
+  switch (layout) {
+    //! NT : K,M x K,N -> K,M,N
+    case MmaLayout::NT:
+      return 0;
+    //! TT : M,K X K,N -> M,K,N
+    case MmaLayout::TT:
+      return 1;
+    //! TN : M,K X N,K -> M,N,K
+    case MmaLayout::TN:
+      return 2;
+    default:
+      TORCH_CHECK(false, "unsupported data layout.");
+  }
+}
+void scheduleSplitKReduction(
```

Comment:
asking some general question: which do you think makes more sense in the long term? Should we build out a separate reduction scheduler for matmul, or should we refactor the reduction scheduler and reuse it for matmul? Do we need different reduction strategies depending on the shape? Does this reduction scheduler need to handle fancy fusions (matmul epilogue fusion)? cc @naoyam @drzejan2 @mmigdal-nv

Reply:
Long term, we should investigate more flexible scheduling/segmentation infrastructure. The fundamental problem is that the current schedulers are monolithic and not composable, so we can't handle cases like this without creating a new combination of patterns, which is obviously not scalable.

Reply:
Also @naoyam, I am not very familiar with reduction scheduling, could you review this part? Also, I have a question, when does a schedule imply cooperative launch, and when it uses semaphores and let the last block to do reduction?

Reply:
> we should investigate more flexible scheduling/segmentation infrastructure

I agree.

Reply:
I'll review the PR too.

In general, the grid reduction should just use semaphores with no cooperative launch. It's the patten of a grid reduction followed by a grid broadcast (aka, allreduce) where the cooperative launch is required. If the reduction part was done without a grid sync, then only the last block would have the valid result, and this value would need to be communicated to all the other blocks, so the other blocks would need to wait for the last block, which means the entire blocks would need to keep resident.

Reply:
> I'll review the PR too.
> 
> In general, the grid reduction should just use semaphores with no cooperative launch. It's the patten of a grid reduction followed by a grid broadcast (aka, allreduce) where the cooperative launch is required. If the reduction part was done without a grid sync, then only the last block would have the valid result, and this value would need to be communicated to all the other blocks, so the other blocks would need to wait for the last block, which means the entire blocks would need to keep resident.

Is it possible to use `allreduce` for reduction, or it is only for persistent kernels?

Reply:
I think it's possible if we want, but we are not using it as it's not necessary, assuming the non-persistent version is more efficient.

Reply:
I haven't looked at the actual PR changes yet (I'll do), I guess the kernel ends up having a grid sync because of inconsistent parallelization with `ParallelType::BID` and `MemoryType::Global`? Think of:

```
tv0_g: [I0, I1]
t1_l: = tv0_g

tv0_g->axis(0)->parallelize(BIDx);
tv0_g->axis(1)->parallelize(BIDy);
tv1_l->axis(0)->parallelize(BIDy);
tv1_l->axis(1)->parallelize(BIDx);
```

In this case we detect a RAW sync is necessary between `tv0` and `tv1`, and since `BID` types are used, a grid sync is inserted. As you can see, this sync needs to be persistent.

Reply:
> Should we build out a separate reduction scheduler for matmul, or should we refactor the reduction scheduler and reuse it for matmul?

That's a good question. In general, having reduction scheduler refactored and used in matmul scheduler makes more sense to me but I'm not familiar with current needs/limitations of existing solution. I wonder if matmul scheduler nature (usage of mma) will not cause in the end that we will have specialization for matmul, and common implementation for other uses.
By separate `reduction scheduler for matmul`, you mean a scheduler that will handle problems with large K and will use mma ops? It will leave separately from current matmul scheduler and by having runtime check we will decide which one to use? (runtime checks are needed, as we need to have an access to problem shape).

> Do we need different reduction strategies depending on the shape?
I don't expect this to happen, although I've seen perf summary which shown that in some cases it is better to use serial split-k (in-place) while in others it is better to use more advanced algorithms. The problem is that architectures of compared kernels were very different so I had to be careful with defining a conclusions. For our need, at this point, I don't expect we would need to have different reductions based on problem shape. It will be a different story with persistent kernels.

> Does this reduction scheduler need to handle fancy fusions (matmul epilogue fusion)?
I think this is the key question. I would expect that for problem with large-k and with relu epilogue we will have a single kernel.